pennylane/ops/qubit/matrix_ops.py

```diff
@@ -332,3 +334,134 @@ def _controlled(self, control):
 
     def label(self, decimals=None, base_label=None, cache=None):
         return super().label(decimals=decimals, base_label=base_label or "U", cache=cache)
+
+
+class BlockEncode(Operation):
+    r"""BlockEncode(a, wires)
+    Apply an arbitrary matrix, :math:`A`, encoded in the top left block of a unitary matrix.
+
+    .. math::
+
+        \begin{align}
+             U(A) &=
+             \begin{bmatrix}
+                A & \sqrt{I-AA^\dagger} \\
+                \sqrt{I-A^\dagger A} & -A^\dagger
+            \end{bmatrix}.
+        \end{align}
+
+    **Details:**
+
+    * Number of wires: Any (the operation can act on any number of wires)
+    * Number of parameters: 1
+    * Number of dimensions per parameter: (2,)
+    * Gradient recipe: None
+
+    Args:
+        A (array[complex]): a general :math:`(n \times m)` matrix to be encoded
+        wires (Sequence[int] or int): the wire(s) the operation acts on
+        do_queue (bool): indicates whether the operator should be
+            recorded when created in a tape context
+        id (str): custom label given to an operator instance,
+            can be useful for some applications where the instance has to be identified
+
+    Raises:
+        ValueError: if the number of wires doesn't fit the dimensions of the matrix
+
+    **Example**
+
+    >>> dev = qml.device('default.qubit', wires=2)
+    >>> A = [[0.1,0.2],[0.3,0.4]]
+    >>> @qml.qnode(dev)
+    ... def example_circuit():
+    ...     qml.BlockEncode(A, wires=range(2))
+    ...     return qml.state()
+    >>> print(qml.matrix(example_circuit)())
+    [[ 0.1         0.2         0.97283788 -0.05988708]
+    [ 0.3         0.4        -0.05988708  0.86395228]
+    [ 0.94561648 -0.07621992 -0.1        -0.3       ]
+    [-0.07621992  0.89117368 -0.2        -0.4       ]]
+    """
```

Comment:
I think its worth adding an example where the matrix A is not square and we need to do some additional padding with identity so that the resulting operator is 2^k x 2^k for some k. 

Maybe something like: 
```
A = [[1, 0, 1],[-1, 1, 0]]  # normalized 
```